### ISSUE TITLE
Onnx Opset=9 Wrong Focus Module( Focus模块的Slice转换出错 ）

### ISSUE THREAD

**Issue by BenjaminLiu1003** [Megvii-BaseDetection/YOLOX#518]
我在使用如下命令进行转换
`python tools/export_onnx.py --output-name yolox_s_opset9 -n yolox-s -c models/yolox_s.pth -o 9`

但是在进行onnx simplify时报错
`
2021-08-18 12:36:35.006 | INFO     | __main__:main:89 - generated onnx model named yolox_s_opset9
[libprotobuf FATAL google/protobuf/stubs/common.cc:61] This program requires version 3.13.0 of the Protocol Buffer runtime library, but the installed version is 2.6.1.  Please update your library.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "google/protobuf/descriptor.pb.cc".)
terminate called after throwing an instance of 'google::protobuf::FatalException'
  what():  This program requires version 3.13.0 of the Protocol Buffer runtime library, but the installed version is 2.6.1.  Please update your library.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "google/protobuf/descriptor.pb.cc".)
Aborted (core dumped)
`

而且得到的未经过simplify的onnx文件也出现了错误
![clipboard-202108171700-rjcoh](https://user-images.githubusercontent.com/57307890/129838087-80a03c93-fa2b-41e0-a120-c2a212df7faf.png)

请问应该如何转换指定opset=9转换成功

**Comment by Demonsu** [comment]
+1

**Comment by MrSuen** [comment]
同样的问题，也许是opset9不支持slice？我尝试在darknet里将stem的focus改成下面dark2-5那样的格式，ksize=3 stride=2，再到你这步的时候依然转换失败......有没有大佬指点一下

**Comment by Demonsu** [comment]
Quick patch see attached

[0001-Fix-onnx-export-opset-9-segmentation-fault-issue.patch.zip](https://github.com/Megvii-BaseDetection/YOLOX/files/7004957/0001-Fix-onnx-export-opset-9-segmentation-fault-issue.patch.zip)

**Comment by sunmooncode** [comment]
@MrSuen  修改network_blocks.py 里面的Focus


class Focus(nn.Module):
    """Focus width and height information into channel space."""

    def __init__(self, in_channels, out_channels, ksize=1, stride=2, act="silu"):
        super().__init__()
        self.conv = BaseConv(in_channels, out_channels, ksize, stride, act=act)

    def forward(self, x):
        # shape of x (b,c,w,h) -> y(b,4c,w/2,h/2)
        # patch_top_left = x[..., ::2, ::2]
        # patch_top_right = x[..., ::2, 1::2]
        # patch_bot_left = x[..., 1::2, ::2]
        # patch_bot_right = x[..., 1::2, 1::2]
        # x = torch.cat(
        #     (
        #         patch_top_left,
        #         patch_bot_left,
        #         patch_top_right,
        #         patch_bot_right,
        #     ),
        #     dim=1,
        # )
        return self.conv(x)

**Comment by BenjaminLiu1003** [comment]
@Demonsu  谢谢回答！这样更改可以转换成功了，但是在用ONNXRuntime里面运行这个模型时，检测不到结果，是需要finetune一下么？

**Comment by BenjaminLiu1003** [comment]
@Demonsu 还是说需要手动做一次Focus里面的Slice操作？

**Comment by BenjaminLiu1003** [comment]
@Demonsu 我把转换的onnx文件的替换Focus的那个Conv_0,Sigmoid还有Mul删除了，直接外部处理Slice拼接image成为( 4 * c ) * ( h / 2 ) * ( w / 2 )的图像输入给Conv_4,然后把OnnxDemo里面的代码相应改好了，但是还是得不到检测结果，请问应该如何处理？谢谢。

**Comment by BenjaminLiu1003** [comment]
@Demonsu 看起来应该只能重新训练了，请问有冻结参数只finetune训练那第一个卷积层的指导么？

谢谢！

**Comment by MrSuen** [comment]
> @MrSuen 修改network_blocks.py 里面的Focus
> 
> class Focus(nn.Module):
> """Focus width and height information into channel space."""
> 
> ```
> def __init__(self, in_channels, out_channels, ksize=1, stride=2, act="silu"):
>     super().__init__()
>     self.conv = BaseConv(in_channels, out_channels, ksize, stride, act=act)
> 
> def forward(self, x):
>     # shape of x (b,c,w,h) -> y(b,4c,w/2,h/2)
>     # patch_top_left = x[..., ::2, ::2]
>     # patch_top_right = x[..., ::2, 1::2]
>     # patch_bot_left = x[..., 1::2, ::2]
>     # patch_bot_right = x[..., 1::2, 1::2]
>     # x = torch.cat(
>     #     (
>     #         patch_top_left,
>     #         patch_bot_left,
>     #         patch_top_right,
>     #         patch_bot_right,
>     #     ),
>     #     dim=1,
>     # )
>     return self.conv(x)
> ```

感谢回答，但是你提供的代码我依然会运行出现问题，我尝试使用楼下发的patch文件，好像是可以的
